src/server/gc_worker/applied_lock_collector.rs

```diff
@@ -0,0 +1,732 @@
+            }
+        }
+    }
+
+    fn load_max_ts(&self) -> TimeStamp {
+        self.state.max_ts.load(Ordering::Acquire).into()
+    }
+}
+
+impl Coprocessor for LockObserver {}
+
+impl QueryObserver for LockObserver {
+    fn pre_apply_query(&self, _: &mut ObserverContext<'_>, requests: &[RaftRequest]) {
+        let max_ts = self.load_max_ts();
+        if max_ts.is_zero() {
+            return;
+        }
+
+        // For each put in CF_LOCK, collect it if it's ts <= max_ts.
+        for req in requests {
+            if req.get_cmd_type() != CmdType::Put {
+                continue;
+            }
+            let put_request = req.get_put();
+            if put_request.get_cf() != CF_LOCK {
+                continue;
+            }
+
+            let lock = match Lock::parse(put_request.get_value()) {
+                Ok(l) => l,
+                Err(e) => {
+                    error!(
+                        "cannot parse lock";
+                        "value" => hex::encode_upper(put_request.get_value()),
+                        "err" => ?e
+                    );
+                    self.send(LockObserverMsg::Err(ErrorInner::Mvcc(e.into()).into()));
+                    return;
+                }
+            };
+
+            if lock.ts <= max_ts {
+                let key = Key::from_encoded_slice(put_request.get_key());
+                self.send(LockObserverMsg::Locks(vec![(key, lock)]));
+            }
+        }
+    }
+}
+
+impl ApplySnapshotObserver for LockObserver {
+    fn pre_apply_plain_kvs(
+        &self,
+        _: &mut ObserverContext<'_>,
+        cf: CfName,
+        kv_pairs: &[(Vec<u8>, Vec<u8>)],
+    ) {
+        if cf != CF_LOCK {
+            return;
+        }
+
+        let max_ts = self.load_max_ts();
+        if max_ts.is_zero() {
+            return;
+        }
+
+        let locks: Result<Vec<_>> = kv_pairs
+            .iter()
+            .map(|(key, value)| {
+                Lock::parse(value)
+                    .map(|lock| (key, lock))
+                    .map_err(|e| ErrorInner::Mvcc(e.into()).into())
+            })
+            .filter(|result| result.is_err() || result.as_ref().unwrap().1.ts <= max_ts)
+            .map(|result| {
+                // `pre_apply_plain_keys` will be invoked with the data_key in RocksDB layer. So we
+                // need to remove the `z` prefix.
+                result.map(|(key, lock)| (Key::from_encoded_slice(origin_key(key)), lock))
+            })
+            .collect();
+
+        match locks {
+            Err(e) => self.send(LockObserverMsg::Err(e)),
+            Ok(l) => self.send(LockObserverMsg::Locks(l)),
+        }
+    }
+
+    fn pre_apply_sst(&self, _: &mut ObserverContext<'_>, cf: CfName, _path: &str) {
+        if cf == CF_LOCK {
+            let e = box_err!("snapshot of lock cf applied from sst file");
+            error!("cannot collect all applied lock"; "err" => ?e);
+            self.send(LockObserverMsg::Err(e));
```

Comment:
Does it mean that green gc won't work when an sst is ingested during observation?

Reply:
Yes. In the current implementation of TiKV, lock cf is always in plain files format when applying snapshot. So I didn't implement it for SST files for now.

Reply:
Ah, I didn't know Lock CF snapshot is using plain format.